docs/product/specs/permissions-revamp.md

```diff
@@ -0,0 +1,115 @@
+  - These privileges should be tested and proper errors and rectification steps should be provided to the admin.
+* Pre-defined roles should be automatically created in this setup.
+
+### Role configuration
+* Admins should be able to configure existing roles in a database server, within a database page.
+  - Wherever possible, we should reflect roles that already exist and let the admin add them by specifying the password.
+* Admins should be able to create new roles on the database server and associate them with the database.
+  - This will require the role used by the admin to have CREATEROLE privilege.
+* Future considerations:
+  - Roles exist on the database server and not under a database.
+  - However, inorder to simplify the user flow, we are not introducing a UX dedicated for database servers.
+  - We can explore this option in the future, based on user feedback or product strategy discussions.
+
+### Access control options
+* Admins should be able to provide Mathesar users access to each database configured in Mathesar.
+* Each Mathesar user, including admins, should only be associated with a single role per database.
+  - Admins should be able to update this user-to-role configuration.
+* All Mathesar admins should be able to access all databases configured in Mathesar. Admins cannot be removed from a database.
+  - When a Mathesar admin is newly added into the product, they will automatically be provided the 'default role' for all databases.
+  - The concept of a 'default role' is explained further below in this document.
+* The role for admins can be changed.
+  - This can be restricted to reduce scope and we could mention that admins will always assume the default role.
+  - The challenges are mentioned in further below in the 'Dealing with Ownership and Collaboration issues' section of this document.
+* Standard Mathesar users can be added/removed from a database.
+* Standard Mathesar users will not be able to access a database without being provided explicit access by a Mathesar admin.
+* The permissions for DB objects will be determined by the priviliges of the role the user is associated with.
+* For Mathesar specific objects, we will have a setting per user per database that allows the user to 'read' or 'read-write'.
+  - This applies for explorations and metadata such as table and column settings.
+  - Future consideration:
+    - We can allow an additional option 'Infer from DB privileges' that will automatically identify access to each Mathesar-object based on the privileges on the underlying DB objects.
+
+### Schema level access control options
+* If a user has ownership access (or appropriate GRANT priviliges) to a schema based on their underlying role, they should be able to grant/revoke access to that schema to other roles configured in Mathesar.
+* While granting access to a role, the user can choose access control at three different levels.
+  - manage: USAGE, GRANT, SELECT, INSERT, UPDATE, DELETE, SET and ALTER
```

Comment:
To be clear, the table from [relevant docs](https://www.postgresql.org/docs/current/ddl-priv.html) is:

| Privilege    | Applicable Object Types                                                        |
|--------------|--------------------------------------------------------------------------------|
| SELECT       | LARGE OBJECT, SEQUENCE, TABLE (and table-like objects), table column           |
| INSERT       | TABLE, table column                                                            |
| UPDATE       | LARGE OBJECT, SEQUENCE, TABLE, table column                                    |
| DELETE       | TABLE                                                                          |
| TRUNCATE     | TABLE                                                                          |
| REFERENCES   | TABLE, table column                                                            |
| TRIGGER      | TABLE                                                                          |
| CREATE       | DATABASE, SCHEMA, TABLESPACE                                                   |
| CONNECT      | DATABASE                                                                       |
| TEMPORARY    | DATABASE                                                                       |
| EXECUTE      | FUNCTION, PROCEDURE                                                            |
| USAGE        | DOMAIN, FOREIGN DATA WRAPPER, FOREIGN SERVER, LANGUAGE, SCHEMA, SEQUENCE, TYPE |
| SET          | PARAMETER                                                                      |
| ALTER SYSTEM | PARAMETER                                                                      |

The spec you have would imply a manager gets:

| Privilege    | Applicable Object Types                                                        |
|--------------|--------------------------------------------------------------------------------|
| SELECT       | LARGE OBJECT, SEQUENCE, TABLE (and table-like objects), table column           |
| INSERT       | TABLE, table column                                                            |
| UPDATE       | LARGE OBJECT, SEQUENCE, TABLE, table column                                    |
| DELETE       | TABLE                                                                          |
| USAGE        | DOMAIN, FOREIGN DATA WRAPPER, FOREIGN SERVER, LANGUAGE, SCHEMA, SEQUENCE, TYPE |
| SET          | PARAMETER                                                                      |

Minor issues:

- There is no `GRANT` privilege. It's an option on each privilege granted.
- There's no need to grant `SELECT` on table columns specifically.
- There's no need to grant anything on table columns specifically at this time.
- There's no need to grant `USAGE` on `FOREIGN DATA WRAPPER`, `FOREIGN SERVER`, or `LANGUAGE`
- I don't think it makes sense to grant `SET` from Mathesar, since we don't use it.
- _There is no ALTER privilege_ !!!

Major problem:

Dropping and altering database objects are privileges exclusive to the owner of the object, and can't be granted except by granting membership in the owning role. This is the main thing that makes me hesitant to try to get schema-level privileges handled by the beta release. Getting that right would mean (at a minimum):

- More emphasis on creating roles exclusively for use as groups, and
- More UI/UX work around this ownership concept.

I don't think we can do it quickly enough.

The same follows for the "Edit" role on schema level. "Viewer" should be possible on a schema level.

Reply:
> hesitant to try to get schema-level privileges handled by the beta

Me too

Reply:
@mathemancer and I had a call to discuss this spec in general and we agreed that manager would directly mean the 'default role', which is also used by the admins.

For any custom role creation, we'd only consider viewer and editor.

Schema-level privileges, however, do require a concept of ownership to make it easier for the user to understand and to accurately depict what's happening in the DB.

We are having another call where we will discuss this more.

Reply:
We've introduced displaying the ownership for objects and there are changes revolving granting of privileges in https://github.com/mathesar-foundation/mathesar-wiki/commit/2ad0d7262212be9b5897deda8b7962e99d2f7191.